### ISSUE TITLE
RFE: allocate static IP's via DHCP

### ISSUE THREAD

**Issue by efenex** [karmab/kcli#11]
It would be nice if we could hand out IP's to VM's through the dnsmasq configuration that is generated by libvirt. This should be possible by using the "host" element under the <ip> block, i.e.:

```
  <ip address='10.0.1.1' netmask='255.255.255.0'>
    <host mac="AA:BB:CC:DD:EE:FF" name="hostname" ip="10.0.1.2">
    <dhcp>
      <range start='10.0.1.2' end='10.0.1.254'/>
    </dhcp>
  </ip>
```

I tried doing this without specifying the mac-address (should be possible cfr. IPv6 capabilities), but that doesn't actually hand out the IP's. 

Benefit of this approach is that hosts are given the correct IP from the moment they boot (they all have DHCP enabled), and don't rely on cloud-init mechanisms to configure/bring up the IP correctly (which they don't always).

**Comment by efenex** [comment]
Implemented this currently by adding the ability to specify a MAC address in a VM definition, and adding the "host" entries specified above in the net definition manually. Not ideal, but it works!

**Comment by karmab** [comment]
i saw your PR and i think RFE makes sense, but  i m gonna implement it differently, as i believe:
- you shouldnt be editing the network on your own
- we should handle this specific case exclusively when user doesnt want/cant  use cloudinit. so setting cloudinit to False will be required

**Comment by efenex** [comment]
Well, I am using cloudinit, just not all of it's functionality (because it doesn't work well) - so I don't think disabling it should be a hard requirement.

As for the editing the network on my own: absolutely agree, that's why I made this RFE, so kcli can create those entries for me ;-)

I did the PR for the MAC address because I thought there may be other use cases where it could be useful/required (i.e. a server that has a license locked to a MAC, or whatever..). I don't really want to set the MAC address myself/statically, just using that as a workaround for now.

**Comment by karmab** [comment]
i ve implemented the logic so that when an ip is provided and cloudinit is disabled, we make the dhcp reservation on the corresponding network retrieving the mac created by libvirt.
i dont know how to combine that with enabled cloudinit actually, cos when cloudinit works properly , it will allow you to force an ip and all the other things too.
Also i believe you can have dhcp properly working, and ips reported in kcli list and available for kcli ssh $IP, and for those scenarios you dont need static ip so your use case is not clear to me

**Comment by efenex** [comment]
well, cloudinit can also do other things besides configuring the network, i.e. the "cmds" array you have with a list of commands to execute post-bootup, adding SSH keys, setting hostname, etc.

In my current setup, I'm using cloud-init to do all of the above, but leaving the network configuration to the DHCP mechanism. 

Of course, it would be better if cloud-init could properly handle the network configuration as well, but there are many situations where it does not. Example: the ubuntu cloud images that rename network interfaces from "eth0" to "ens3" or similar. In this case, you need to know what your interface name will be (what ubuntu makes of it) in advance, in order to be able to specify a static IP through cloud-init. Using the DHCP mechanism, you don't need to know.

With the current kcli code (last I checked), a static network configuration for cloud-init is only generated  when a "mask" parameter is specified, otherwise your interface ends up being DHCP. This means that the following configuration:

nets:
- name: internal1
  ip: 10.0.0.10

Does not work as expected: i.e. the IP is not allocated to the server, but a DHCP IP is given to the VM. So in my opinion, detecting a situation like this (DHCP enabled, cloud-init enabled, IP given but no mask specified) would actually already be a good use-case for registering hosts in the virsh net with MAC/IP.


Alternatively, an explicit option could be added, i.e.:

nets:
- name: internal1
  ip: 10.0.0.10
  reserveip: true

Default for "reserveip" would be false, in which case everything can continue to work exactly as it does now. But when reserveip is set to true, it could configure the "host" entry in the DHCP network regardless of the fact whether cloud-init is enabled or disabled. This makes the configuration explicit, and the user aware of potential pitfalls.

**Comment by karmab** [comment]
Yeah we will go for the specific option at net level, i was just beeing
lazy but you ve got a point with nic os renaming ( which is crap btw)

**Comment by karmab** [comment]
i ve added the keyword reserveip at profile level and within plan to force a dhcp reservation.
let me know if that works for you and we will close this issue.
btw i ve also added the network creation in plan files so that one can describe completely a deployment ( and create only missing resources ). Might also add a disks section for shared disks

**Comment by efenex** [comment]
Awesome, thanks! Will test and get back to you. I'll close the issue already as I'll be away for a bit and don't know when I can test thoroughly - but I'll get you feedback once I do!

**Comment by efenex** [comment]
Karim,

I've finally gotten to redeploying our box and testing all the new functionality in kcli. I'm happy to report that almost everything I want/use works out of the box now, with only one minor issue left.

When we deploy a new VM, using cloudinit = true and reserveip = true, the cloud-init template that is pushed by kcli still includes a network definition. That means that in this case, the DHCP functionality is not used, and the IP is statically configured by cloud-init.

Our use-case is:

- I want to use cloud-init, but not for network (so to execute some bootstrap commands, set hostname, etc)
- I want to use the DHCP mechanism for networking, as this is guaranteed to work with every distribution, cloud-init or not

In order to achieve this behavior, I changed the following line (line number 828) in the _cloudinit function in __init__.py from:

if ip is not None and netmask is not None:

changed to:

if ip is not None and netmask is not None and not reserveip:


Is this something you would accept upstream, and would you like me to make a PR for it?

**Comment by karmab** [comment]
Dont bother, i ll make the change tonight ;)
Btw, i ve dockerized the install and plan to add basic docker support to
the plans

Regards,

**Comment by karmab** [comment]
i ve made the indicated change and published a 2.17 pypi package.
while not related to this issue, note still that kcli isnt heavily tested with lvm based pools :(
